ziggurat-core-geoip/src/geoip.rs

```diff
@@ -0,0 +1,130 @@
+//use ipinfo::{IpInfo, IpInfoConfig};   // TODO(asmie): implement IpInfo provider
+use ip2location::{Record, DB};
+use ipgeolocate::{Locator, Service};
+
+/// List of supported geoip providers.
+pub enum Provider {
+    IpApiCom,
+    IpApiCo,
+    IpInfo,
+    Ip2Location,
+}
+
+/// Geoip provider service configuration.
+pub struct GeoIPService {
+    /// Geoip provider. Look at `Provider` enum for supported providers.
+    pub provider: Provider,
+    /// API key (paid or free) for online providers. For offline providers put here path to database file.
+    pub api_key: String,
+}
+
+/// IP information
+pub struct IPGeoInfo {
+    /// IP address
+    pub ip: String,
+    /// Country name (long name)
+    pub country: Option<String>,
+    /// City name
+    pub city: Option<String>,
+    /// IP address latitude
+    pub latitude: Option<f64>,
+    /// IP address longitude
+    pub longitude: Option<f64>,
+    /// Timezone of the IP
+    pub timezone: Option<String>,
+    /// ISP name (unavailable for some providers)
+    pub isp: Option<String>,
+}
+
+impl GeoIPService {
+    pub fn new(provider: Provider, api_key: String) -> Self {
+        Self { provider, api_key }
+    }
+
+    pub async fn lookup_ip(&self, ip: String) -> Result<IPGeoInfo, String> {
+        match self.provider {
+            Provider::IpInfo => {
+                unimplemented!("IpInfo not implemented yet");
+            }
+            Provider::IpApiCom | Provider::IpApiCo => {
+                let service = match self.provider {
+                    Provider::IpApiCom => Service::IpApi,
+                    Provider::IpApiCo => Service::IpApiCo,
+                    _ => unreachable!(),
+                };
+
+                match Locator::get(ip.as_str(), service).await {
+                    Ok(loc_ip) => Ok(IPGeoInfo {
+                        ip,
+                        country: Some(loc_ip.country),
+                        city: Some(loc_ip.city),
+                        latitude: match loc_ip.latitude.parse::<f64>() {
+                            Ok(lat) => Some(lat),
+                            Err(_) => None,
+                        },
+                        longitude: match loc_ip.longitude.parse::<f64>() {
+                            Ok(long) => Some(long),
+                            Err(_) => None,
+                        },
+                        timezone: Some(loc_ip.timezone),
+                        isp: Some("".to_string()),
```

Comment:
```suggestion
                        isp: Some("".to_owned()),
```
super-nit: I prefer `to_owned` to `to_string` with `&str` in order to be able to find actual conversions to `String` easier (if need be)